[CC-25831] Document the Push Notification feature (#1590)
spryker/spryker-docs: docs/scos/user/features/202304.0/push-notification-feature-overview.md

---
title: Push Notification feature overview
description: The Push Notifications feature lets you receive push notifications
last_updated: Feb 2, 2023
template: concept-topic-template
---

The *Push Notification* feature lets users subscribe to the web push notifications, which are sent from the server to all registered subscriptions.

{% info_block infoBox "Note" %}

The feature can be integrated into the non-Srpyker customer application only. Spryker features do not support push notifications.

{% endinfo_block %}

> ## Subscribing to notifications

Reviewer: This part should have the information about generating and storing the public key which is later used to identify the browser to send notifications to

Author: I'll try to come up with something. I'll move it out to a separate chapter though

> There are two tasks that enable the feature. The first lets the user subscribe to the web push notifications. For this, the site page needs a service worker with which the user is registered on the server and received messages are processed. The service worker is downloaded to the user’s platform in the background, which lets it be executed outside of the site page's context.

Reviewer: What's the second task?

Author: That was written based on the primary source. However, even the primary source does not have the second task mentioned. I'll get rid of that part.

> If a user subscribes to the service, the service worker is registered. The service worker requests all required information through the web push API and sends this through an HTTP request to the server. The server stores this information in its database to send notifications to the client.

Reviewer: ```suggestion
After a user subscribes to the service, the service worker is registered. The service worker requests all required information through the web push API and sends this through an HTTP request to the server. The server stores this information in its database to send notifications to the client.
```

## Sending notifications

> Notifications are sent from the server to all registered subscriptions with an HTTP request. To correctly identify a user, a signature must be transmitted in the request header. This signature is generated from the public key used for registration and the private key. The actual message and further information are transmitted as user data. If the encryption and formatting are correct and the signature is validated, the push service sends the notification to the user.

Reviewer: It's not clear what `further information` is.

Reviewer: ```suggestion
Notifications are sent from the server to all registered subscriptions with an HTTP request. To correctly identify a user, a signature must be transmitted in the request header. This signature is generated from the public key used for registration and the private key. The actual message and further information are transmitted as user data. If the encryption and formatting are correct, and the signature is validated, the push service sends the notification to the user.
```

<script type="text/javascript" src="https://viewer.diagrams.net/js/viewer-static.min.js"></script>

1. A user opens a page that wants to send push notifications. The website installs a service worker, which is permanently saved by the browser and used later to display push notifications.
2. The website displays an authorization prompt to the user.

> 3. If the user accepts to receive push notifications, the browser contacts its own push service—for example, Mozilla Push Service or Google Firebase Cloud Messaging (FCM)—to create a new mailbox, a secret URL called an endpoint. Anyone who knows this URL can send push notifications to that browser.

Reviewer: ```suggestion
3. If the user accepts to receive push notifications, the browser contacts its own push service—for example, Mozilla Push Service or Google Firebase Cloud Messaging (FCM)—to create a new mailbox and a secret URL called an endpoint. Anyone who knows this URL can send push notifications to that browser.
```

Reviewer: Otherwise, I don't get the meaning of that part of the sentence

> 4. The browser gives the endpoint to the web page through a Javascript callback, and the endpoint is sent to the Spryker Cloud Commerce Operation System (SCCOS) application server for storage.

Reviewer: ```suggestion
4. The browser gives the endpoint to the web page through a Javascript callback, and the endpoint is sent to the Spryker Cloud Commerce OS (SCCOS) application server for storage.
```

Reviewer: Also, not operation system, but an operating system :D

---
title: Push Notification feature overview
description: The Push Notifications feature lets you receive push notifications
last_updated: Feb 2, 2023
template: concept-topic-template
---

The *Push Notification* feature lets users subscribe to the web push notifications, which are sent from the server to all registered subscriptions.

{% info_block infoBox "Note" %}

The feature can be integrated into the non-Srpyker customer application only. Spryker features do not support push notifications.

{% endinfo_block %}

## Key pair

> To set up the process, you need to generate a private and public key pair. These two keys are generated once for a website and are known as *Voluntary Application Server Identification for Web Push (VAPID)* keys. The private key must be hidden from the end user and must only be used on the server side to encrypt created notifications. The utilization of VAPID protection guarantees that only authorized servers are permitted to send notifications to clients

Author: ```suggestion
To set up the process, you need to generate a private and public key pair. These two keys are generated once for a website and are known as *Voluntary Application Server Identification for Web Push (VAPID)* keys. The private key must be hidden from the end user and must only be used on the server side to encrypt created notifications. The utilization of VAPID protection guarantees that only authorized servers are permitted to send notifications to clients.
```
